server/src/Api/Handlers.hs

```diff
@@ -46,6 +67,57 @@ hashScript :: HashScriptRequest -> AppM HashedScript
 hashScript (HashScriptRequest script) =
   pure . HashedScript $ hashLedgerScript script
 
+finalizeTx :: FinalizeRequest -> AppM FinalizedTransaction
+finalizeTx req@(FinalizeRequest {tx, datums, redeemers}) = do
+  pparams <- asks protocolParams
+  liftIO $ do
+    putStrLn "cbor:"
+    print req
+  decodedTx <- maybe (handleError $ InvalidHex "Failed to decode Tx") pure $
+    decodeCborValidatedTx tx
+  decodedRedeemers <- maybe (handleError $ InvalidHex "Failed to decode Redeemers") pure $
+    decodeCborRedeemers redeemers
+  decodedDatums <- maybe (handleError $ InvalidHex "Failed to decode Datums") pure $
+    traverse decodeCborDatum datums
+  let
+    languages = Set.fromList [PlutusV1, PlutusV2]
+    txDatums = TxWitness.TxDats . Map.fromList $
+      decodedDatums <&> \datum -> (Data.hashData datum, datum)
+    mbIntegrityHash = Tx.hashScriptIntegrity
+      (C.toLedgerPParams C.ShelleyBasedEraAlonzo pparams)
+      languages
+      decodedRedeemers
+      txDatums
+  liftIO $ do
+    putStrLn "tx:"
+    print decodedTx
+    putStrLn "redeemres:"
+    print decodedRedeemers
+    putStrLn "datums:"
+    print decodedDatums
+    putStrLn "integrity hash:"
+    print mbIntegrityHash
+  let
+    addIntegrityHash t =
+      t { body = body t &
+          \body -> body { scriptIntegrityHash = mbIntegrityHash } }
+    addDatumsAndRedeemers t =
+      t { wits = wits t &
+          \witness -> witness { txdats = txDatums, txrdmrs = decodedRedeemers } }
+    finalizedTx = addIntegrityHash $ addDatumsAndRedeemers decodedTx
+    response = FinalizedTransaction . encodeCborText . Cbor.serializeEncoding $
+      Tx.toCBORForMempoolSubmission finalizedTx
```

Comment:
Hopefully this method of serializing to CBOR will work with Nami. Just making a note here for later debugging (if it comes up of course)

Reply:
It is working with `serialization-lib`, so there's a good chance it'll work with Nami.